hibernate-core/src/main/java/org/hibernate/graph/internal/parse/GraphParser.java

```diff
@@ -0,0 +1,167 @@
+	public static final Logger PARSING_LOGGER = Logger.getLogger( "org.hibernate.orm.graph.parsing" );
+
+	/**
+	 * Parse the passed graph textual representation into the passed Graph.
+	 */
+	public static void parseInto(
+			GraphImplementor<?> targetGraph,
+			CharSequence graphString,
+			SessionFactoryImplementor sessionFactory) {
+		final GraphParser instance = new GraphParser( graphString, sessionFactory );
+		instance.graphStack.push( targetGraph );
+		try {
+			instance.graph();
+		}
+		catch (RecognitionException | TokenStreamException e) {
+			throw new InvalidGraphException( "Error parsing graph string" );
+		}
+	}
+
+	private final SessionFactoryImplementor sessionFactory;
+
+	private final Stack<GraphImplementor<?>> graphStack = new StandardStack<>();
+	private final Stack<AttributeNodeImplementor<?>> attributeNodeStack = new StandardStack<>();
+	private final Stack<SubGraphGenerator> graphSourceStack = new StandardStack<>();
+
+	private GraphParser(CharSequence charSequence, SessionFactoryImplementor sessionFactory) {
+		super( new GraphLexer( new CharSequenceReader( charSequence ) ) );
+		this.sessionFactory = sessionFactory;
+	}
+
+
+	@Override
+	protected void startAttribute(Token attributeNameToken) {
+		final String attributeName = attributeNameToken.getText();
+
+		if ( PARSING_LOGGER.isDebugEnabled() ) {
+			PARSING_LOGGER.debugf(
+					"%s Start attribute : %s",
+					StringHelper.repeat( ">>", attributeNodeStack.depth() + 1 ),
+					attributeName
+			);
+		}
+
+		final AttributeNodeImplementor attributeNode = resolveAttributeNode( attributeName );
+		attributeNodeStack.push( attributeNode );
+
+		graphSourceStack.push( PathQualifierType.VALUE.getSubGraphCreator() );
+	}
+
+	private AttributeNodeImplementor resolveAttributeNode(String attributeName) {
+		final GraphImplementor<?> currentGraph = graphStack.getCurrent();
+		assert currentGraph != null;
+
+		final AttributeNodeImplementor attributeNode = currentGraph.addAttributeNode( attributeName );
+		assert attributeNode != null;
+
+		return attributeNode;
+	}
+
+	@Override
+	protected void startQualifiedAttribute(Token attributeNameToken, Token qualifierToken) {
+		final String attributeName = attributeNameToken.getText();
+		final String qualifierName = qualifierToken.getText();
+
+		if ( PARSING_LOGGER.isDebugEnabled() ) {
+			PARSING_LOGGER.debugf(
+					"%s Start qualified attribute : %s.%s",
+					StringHelper.repeat( ">>", attributeNodeStack.depth() + 1 ),
+					attributeName,
+					qualifierName
+			);
+		}
+
+		final AttributeNodeImplementor<?> attributeNode = resolveAttributeNode( attributeName );
+		attributeNodeStack.push( attributeNode );
+
+		final PathQualifierType pathQualifierType = resolvePathQualifier( qualifierName );
+
+		graphSourceStack.push( pathQualifierType.getSubGraphCreator() );
+	}
+
+	private PathQualifierType resolvePathQualifier(String qualifier) {
+		if ( "key".equalsIgnoreCase( qualifier ) ) {
```

Comment:
Why ignoreCase?

Reply:
1. Opens up freedom that would have to be maintained forever after :)
2. Java is case sensitive, not case insensitive. key, KEY, kEy, ... are all different.
3. Java calls these precisely as you spelled them out - lowercase.
4. Should not be an issue here but technically "ignoreCase" is locale-specific and can yield different results in different locales. I am not aware of characters in "key" and "value" that would suffer from this but I am aware of others. For example, `"i".equalsIgnoreCase("I")` is not always `true`.
5. Opens up variation that people will take, resulting in differing styles.
6. Although extremely unlikely, at some point in time in the future there may pop up a need for "KEY" or "Key" or similar with a different meaning but it would have been taken.
7. Completely unimportant but for the sake of foolish completeness, it is a bit slower and you took longer to type it :)

Reply:
If it's supposed to be some sort of keyword, I think I would make it case sensitive too.

Reply:
1. Um, ok.  Its not any different than we do for HQL or (I think) 
2. 

> 1. Opens up freedom that would have to be maintained forever after :)

Um, ok :) Its not any different than we do for HQL or (I think) other "paths" that end in key, index, indices, element, elements, values...  They all accept them case insensitively.

> 2. Java is case sensitive, not case insensitive. key, KEY, kEy, ... are all different.

This depends how you view "key".  You interpret it as an identifier.  And yes Java identifiers are case-sensitive.  However, JPA really defines this as a keyword (look at its keyword list) and as such it ought to be recognized case-insensitively.

> 3. Java calls these precisely as you spelled them out - lowercase.

Well, *Java* does not define "key" as anything.  You mean JPA?  Right, we should never support anything beyond JPA.  Even stuff like parsing "entity graph strings" ;)  Plus, see (2)


> 4. Should not be an issue here but technically "ignoreCase" is locale-specific and can yield different results in different locales. I am not aware of characters in "key" and "value" that would suffer from this but I am aware of others. For example, `"i".equalsIgnoreCase("I")` is not always `true`.

This is not true.  equalsIgnoreCase performs locale-independent checking

> 5. Opens up variation that people will take, resulting in differing styles.

So consistently just like HQL, etc?  I'll take it! :)

> 6. Although extremely unlikely, at some point in time in the future there may pop up a need for "KEY" or "Key" or similar with a different meaning but it would have been taken.

^^ reach ;)

> 7. Completely unimportant but for the sake of foolish completeness, it is a bit slower and you took longer to type it :)

Its slower because it does TheRightThing wrt case-sensitivity.

Reply:
> If it's supposed to be some sort of keyword, I think I would make it case sensitive too.

"select" is also a keyword in JPA.  So you disagree with JPA that "SELECT", "select", "SeLeCt", .. should all be the same?  I do not agree with that.  KEY is another JPA keyword - they should all behave consistently.

Reply:
BTW - section _4.4.1 Identifiers_ is the one that details JPA's key/reserved words and its case-insensitive expectations for their use.

> The following are reserved identifiers: ..., KEY, ... Reserved identifiers are case insensitive.